scipy/signal/_signaltools.py

```diff
@@ -1297,6 +1300,54 @@ def choose_conv_method(in1, in2, mode='full', measure=False):
     return 'direct'
 
 
+def convolve_dispatcher(in1, in2, mode='full', method='auto'):
+    """ Deduce the array namespace and if we can dispatch to the cupy namesake.
+
+    The signature must agree with that of convolve itself.
+
+    NB: Currently, raise if the namespace is CuPy and cannot dispatch to it.
+    """
+    xp = array_namespace(in1, in2)
+    if is_cupy(xp):
+        # Conditions are accurate as of CuPy 13.x
+        if method == "auto":
+            can_dispatch = in1.ndim == 1 and in2.ndim == 1
+        else:
+            can_dispatch = True
+
+        # XXX: inputs are cupy arrays and cupyx.scipy.convolve cannot be used
+        # do we want to convert to numpy/use the CPU version/convert back instead?
+        if not can_dispatch:
+            raise ValueError(
+                f"Cannot dispatch to CuPy for {in1.ndim = }, {in2.ndim = }, "
+                f"{mode = } and {method = }."
+            )
+    else:
+        can_dispatch = False
+
+    return xp, can_dispatch
+
+
+def dispatch_cupy(dispatcher, module_name):
+    def inner(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwds):
+            xp, can_dispatch = dispatcher(*args, **kwds)
+
+            if can_dispatch:
+                # delegate to the cupyx namesake
+                import importlib
+                cupyx_module = importlib.import_module(f"cupyx.scipy.{module_name}")
+                cupyx_func = getattr(cupyx_module, func.__name__)
+                return cupyx_func(*args, **kwds)
```

Comment:
If you want to try extending this support to other libraries which have a `scipy` namespace, consider using [`scipy_namespace_for(xp)`](https://github.com/scipy/scipy/blob/b421cd64d98c16811f84efbfab7701b335f811be/scipy/_lib/_array_api.py#L434-L452). That only has JAX in addition right now (since there is no `torch.scipy`), but it can still help reduce the LOC for dispatching a little, e.g. [`get_array_special_func`](https://github.com/scipy/scipy/blob/b421cd64d98c16811f84efbfab7701b335f811be/scipy/special/_support_alternative_backends.py#L21-L32).

Reply:
I'm currently GPU-less, but have you tried `getattr(xps.signal, func.__name__)`? That's pretty much what we do in `special` right now - is that broken? Running the `special` alternative backends tests on CuPy should show if it is broken, no?

Reply:
```
In [1]: import cupyx

In [2]: import cupyx.scipy

In [3]: getattr(cupyx.scipy, 'signal.convolve')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 getattr(cupyx.scipy, 'signal.convolve')

AttributeError: module 'cupyx.scipy' has no attribute 'signal.convolve'
```

Reply:
The example in special is:

```
import cupyx
getattr(cupyx.scipy.special, gammainc, None)
```

What happens if you run that? Is it different if you run `python dev.py test -b copy -t scipy.special.tests.test_support_alternative_backends`?

Reply:
`special` seems special in cupy:

```
In [1]: import cupyx
   ...: getattr(cupyx.scipy.special, 'gammainc', None)
Out[1]: <ufunc 'cupyx_scipy_gammainc'>

In [2]: import cupyx
   ...: getattr(cupyx.scipy.signal, 'convolve', None)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 2
      1 import cupyx
----> 2 getattr(cupyx.scipy.signal, 'convolve', None)

AttributeError: module 'cupyx.scipy' has no attribute 'signal'

In [3]: import cupyx
   ...: getattr(cupyx.scipy.linalg, 'solve', None)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[3], line 2
      1 import cupyx
----> 2 getattr(cupyx.scipy.linalg, 'solve', None)

AttributeError: module 'cupyx.scipy' has no attribute 'linalg'

In [4]: import cupyx
   ...: getattr(cupyx.scipy.ndimage, 'correlate', None)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 2
      1 import cupyx
----> 2 getattr(cupyx.scipy.ndimage, 'correlate', None)

AttributeError: module 'cupyx.scipy' has no attribute 'ndimage'
```

and indeed `$ python dev.py test -b cupy -t scipy.special.tests.test_support_alternative_backends` is green for me locally.

Reply:
So can we just add the minimal tweak to make `special` less special? I'd really like the mechanism of calling out to the `scipy.submodule` namespace in another library to not be so different between various submodules. It's be better to have a file `_support_alternative_backends.py` also in `signal`, and have the same kind of exposing of it in `__init__.py` rather than with a decorator.

Reply:
Indeed, it sounds like @kmaehashi is open to making `cupyx.scipy` behave like `scipy` w.r.t imports in cupy/cupy#8336